File: benchmark/template/programs/swift/micro/mb11.swift
// MB-11 - Collections: Swift's own Dictionary, Set and Array under
// insert / update / lookup / delete / iterate. Default capacity everywhere.
// Section 4.12(d) pins Dictionary<Int,Int>, Set<Int> and Array<Int> for swift.
#if canImport(Darwin)
import Darwin  // fflush(stdout) for the section 5.2 ITER lines
#else
import Glibc   // the same fflush(stdout) on Linux, where the benchmark runs
#endif

/// Lehmer / MINSTD generator, frozen for every language in the suite.
struct Lcg {
    private var state: Int

    init(seed: Int) {
        state = seed
    }

    mutating func nextInt() -> Int {
        state = (48271 * state) % 2147483647
        return state
    }
}

/// The whole workload body, generation included (methodology 06 section 4, MB-11).
/// Section 5.2 requires each steady iteration to re-seed and regenerate `keys`.
func workload() -> String {
    let n = 1000000
    let r = 3
    let km = 500009
    let sm = 100003
    let q = 1000000007

    var gen = Lcg(seed: 20271917)
    var keys = [Int](repeating: 0, count: n)
    for i in 0..<n {
        keys[i] = gen.nextInt()
    }

    var acc = 0
    var size1 = 0
    var found = 0
    var vsum = 0
    var mchk = 0
    var size2 = 0
    var size3 = 0
    var lsum = 0

    for round in 0..<r {
        keys[round] = keys[round] + 1000000  // anti-elimination, stays below 2^31

        var m = [Int: Int]()
        for i in 0..<n {
            let k = keys[i] % km
            m[k] = (m[k] ?? 0) + 1
        }
        size1 = m.count

        found = 0
        vsum = 0
        for i in 0..<n {
            let k = (keys[i] + 7) % km
            if let v = m[k] {
                found += 1
                vsum += v
            }
        }

        mchk = 0
        for (k, v) in m {  // commutative, so map iteration order does not matter
            mchk = (mchk + (k % 1000003) * v) % 1000003
        }

        for i in stride(from: 0, to: n, by: 2) {
            m.removeValue(forKey: keys[i] % km)
        }
        size2 = m.count

        var st = Set<Int>()
        for i in 0..<n {
            st.insert(keys[i] % sm)
        }
        size3 = st.count

        var lst = [Int]()
        for i in 0..<n {
            lst.append(keys[i] % 1000)
        }
        lsum = 0
        for v in lst {
            lsum += v
        }

        for value in [size1, found, vsum, mchk, size2, size3, lsum] {
            acc = (acc * 31 + value) % q
        }
    }

    return "MB11 \(acc) \(size1) \(found) \(vsum) \(mchk) \(size2) \(size3) \(lsum)"
}

// Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
func main() {
    let args = CommandLine.arguments
    let mode = args.count > 1 ? args[1] : "once"
    var line: String
    if mode == "steady" {
        var iterations = args.count > 2 ? (Int(args[2]) ?? 7) : 7
        if iterations < 1 {
            iterations = 7
        }
        let clock = ContinuousClock()
        line = ""
        for k in 0..<iterations {
            let elapsed = clock.measure {
                line = workload()
            }
            let parts = elapsed.components
            let ns = parts.seconds * 1_000_000_000 + parts.attoseconds / 1_000_000_000
            print("ITER \(k) \(ns)")
            fflush(stdout)  // section 5.2: each ITER line is flushed immediately
        }
    } else {
        line = workload()
    }
    print(line)
}

main()

File: benchmark/template/programs/kotlin/micro/mb11.kt
// MB-11 - Collections: standard hash map, hash set and growable array under load.

/** Lehmer / MINSTD generator, frozen for every language in the suite. */
class Lcg(private var state: Long) {
    fun nextInt(): Long {
        state = (48271L * state) % 2147483647L
        return state
    }
}

const val SEED = 20271917L
const val N = 1_000_000
const val R = 3
const val KM = 500009L
const val SM = 100003L
const val Q = 1_000_000_007L

class Mb11Result(
    val acc: Long,
    val size1: Long,
    val found: Long,
    val vsum: Long,
    val mchk: Long,
    val size2: Long,
    val size3: Long,
    val lsum: Long
)

fun workload(): Mb11Result {
    // Section 5.2: the body re-seeds the generator and regenerates `keys`, so
    // every steady iteration performs exactly the work `once` performs.
    val gen = Lcg(SEED)
    val keys = LongArray(N)
    for (i in 0 until N) {
        keys[i] = gen.nextInt()
    }

    var acc = 0L
    var size1 = 0L
    var found = 0L
    var vsum = 0L
    var mchk = 0L
    var size2 = 0L
    var size3 = 0L
    var lsum = 0L

    for (r in 0 until R) {
        keys[r] = keys[r] + 1000000 // anti-elimination, part of the frozen algorithm

        val m = HashMap<Long, Long>()
        for (i in 0 until N) {
            val k = keys[i] % KM
            m[k] = (m[k] ?: 0L) + 1
        }
        size1 = m.size.toLong()

        found = 0L
        vsum = 0L
        for (i in 0 until N) {
            val k = (keys[i] + 7) % KM
            val v = m[k]
            if (v != null) {
                found++
                vsum += v
            }
        }

        mchk = 0L
        for ((k, v) in m) {
            mchk = (mchk + (k % 1000003) * v) % 1000003
        }

        for (i in 0 until N step 2) {
            val k = keys[i] % KM
            m.remove(k)
        }
        size2 = m.size.toLong()

        val st = HashSet<Long>()
        for (i in 0 until N) {
            st.add(keys[i] % SM)
        }
        size3 = st.size.toLong()

        val lst = ArrayList<Long>()
        for (i in 0 until N) {
            lst.add(keys[i] % 1000)
        }
        lsum = 0L
        for (v in lst) {
            lsum += v
        }

        for (value in longArrayOf(size1, found, vsum, mchk, size2, size3, lsum)) {
            acc = (acc * 31 + value) % Q
        }
    }

    return Mb11Result(acc, size1, found, vsum, mchk, size2, size3, lsum)
}

fun main(args: Array<String>) {
    // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
    val mode = if (args.isNotEmpty()) args[0] else "once"
    var res: Mb11Result
    if (mode == "steady") {
        val k = if (args.size > 1) args[1].toIntOrNull() ?: 7 else 7
        val iterations = if (k < 1) 7 else k
        res = Mb11Result(0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L)
        for (i in 0 until iterations) {
            val t0 = System.nanoTime() // section 5.2 pins System.nanoTime() for Kotlin
            res = workload()
            val elapsed = System.nanoTime() - t0
            println("ITER $i $elapsed")
            System.out.flush() // section 5.2: each ITER line is flushed immediately
        }
    } else {
        res = workload()
    }
    println(
        "MB11 ${res.acc} ${res.size1} ${res.found} ${res.vsum} ${res.mchk} " +
            "${res.size2} ${res.size3} ${res.lsum}"
    )
}
